include/deal.II/differentiation/ad/ad_number_traits.h

```diff
@@ -1060,58 +1082,365 @@ namespace Differentiation
+       * taken with this (floating point) number
+       */
+      static constexpr unsigned int n_supported_derivative_levels = 0;
+
+#  endif
+
+
       /**
        * Underlying floating point value type.
        * This could real-valued or complex-valued.
        */
       using scalar_type = ScalarType;
 
-      static ScalarType
-      get_directional_derivative(const ScalarType & /*x*/,
-                                 const unsigned int /*direction*/)
+
+      /**
+       * Type for real numbers
+       */
+      using real_type =
+        typename internal::RemoveComplexWrapper<ScalarType>::type;
+
+
+      /**
+       * Type for complex numbers
+       */
+      using complex_type = std::complex<real_type>;
+
+
+      /**
+       * The actual auto-differentiable number type
+       */
+      using ad_type = typename std::
+        conditional<is_real_valued, real_type, complex_type>::type;
+
+      /**
+       * The actual auto-differentiable number directional derivative type
+       */
+      using derivative_type = typename std::
+        conditional<is_real_valued, real_type, complex_type>::type;
+
+
+      /**
+       * Extract the value of an auto-differentiable number
+       */
+      static scalar_type
+      get_scalar_value(const ad_type &x)
+      {
+        return x;
+      }
+
+
+      /**
+       * Extract the derivative value of an auto-differentiable number
+       */
+      static derivative_type
+      get_directional_derivative(const ad_type &, const unsigned int)
+      {
+        AssertThrow(
+          false,
+          ExcMessage(
+            "Floating point / arithmetic numbers have no directional derivatives."));
+        return derivative_type();
+      }
+
+
+      /**
+       * Extract the number of directional derivatives value tracked by
+       * an auto-differentiable number
+       */
+      static unsigned int
+      n_directional_derivatives(const ad_type &)
       {
-        // If the AD drivers are correctly implemented then we should not get
-        // here. This is essentially a dummy for when the ADNumberTypeCode for
-        // the original AD number (from which one is getting a derivative >= 2)
-        // is one that specified ADOL-C taped and tapeless numbers, or a
-        // non-nested Sacado number.
         AssertThrow(
           false,
           ExcMessage(
-            "Floating point numbers have no directional derivatives."));
-        return 0.0;
+            "Floating point / arithmetic numbers have no directional derivatives."));
+        return 0;
       }
+
+
+      static_assert((is_real_valued == true ?
+                       std::is_same<ad_type, real_type>::value :
+                       std::is_same<ad_type, complex_type>::value),
+                    "Incorrect template type selected for ad_type");
+
+      static_assert((is_complex_valued == true ?
+                       boost::is_complex<scalar_type>::value :
+                       true),
```

Comment:
I don't understand this either. If `is_complex_valued` is `true`, then `boost::is_complex<scalar_type>::value` should also be true. So the whole condition is true.